### PR TITLE
Bug 1436195 - implement exposing services publicly

### PR DESCRIPTION
**Pull request by djmitche**
This introduces a new package designed to expose services available on
the worker's loopback interface to the public Internet.  The services
can include both HTTP services, which are proxied directly, and TCP
services, which are made available via binary messages on a websocket.

### CODE REVIEW

**Comment by djmitche** [comment]
Pete, could you give this a quick look before I proceed?  I don't want to go too far if this is a totally bogus idea.

My plans from here are:

 * implement an Exposer interface that uses statelessdns
 * modify livelog to use Exposer
 * implement an Exposer interface that uses websocktunnel

This leaves ample opportunity to proxy RDP over WS, and/or support the docker-worker mode of proxying VNC and terminal over WS, when those features are implemented.

I've written automated tests for the important bits here, but also validated this using a bunch of short test programs and various http, websocket and tcp clients and servers.

File: expose/local/local_test.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	exposer := makeExposer(t)
+	exposure, err := exposer.ExposeHTTP(uint16(testPort))
+	if err != nil {
+		t.Fatalf("ExposeHTTP returned an error: %v", err)
+	}
+	defer exposure.Close()
+
+	gotURL := exposure.GetURL()
+	host, port, _ := net.SplitHostPort(gotURL.Host)
+	assert.Equal(t, "http", gotURL.Scheme, "Should return URL with correct schem")
```
Comment (petemoore) [inline expose/local/local_test.go]: scheme

File: expose/local/local_test.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	res, err := http.Get(gotURL.String())
+	if err != nil {
+		t.Fatal(err)
+	}
+	greeting, err := ioutil.ReadAll(res.Body)
+	res.Body.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+	assert.Equal(t, "Hello, world", string(greeting), "got greeting via proxy")
```
Comment (petemoore) [inline expose/local/local_test.go]: Let's also assert a 200 status code. Perhaps a non-200 status code test would also be good, both with and without a response body.

Comment (djmitche) [reply expose/local/local_test.go]: I'll add that to proxy_test.go.

**Comment by djmitche** [comment]
I'll open a new PR when the next few steps are done.

**Pull request closed without merging**